prometheus/graphite/bridge.go

```diff
@@ -0,0 +1,230 @@
+// Copyright 2016 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package graphite provides a bridge to push Prometheus metrics to a Graphite
+// server.
+package graphite
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"regexp"
+	"sort"
+	"strings"
+	"time"
+
+	"golang.org/x/net/context"
+
+	dto "github.com/prometheus/client_model/go"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/expfmt"
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/common/model"
+)
+
+// Config defines the graphite bridge config.
+type Config struct {
+	// The url to push data to. Required.
+	URL string
+
+	// The interval to use for pushing data to Graphite. Defaults to 15 seconds.
+	Interval time.Duration
+
+	// The Gatherer to use for metrics. Defaults to prometheus.DefaultGatherer.
+	Gatherer prometheus.Gatherer
+
+	// The logger that messages are written to. Defaults to log.Base().
+	Logger log.Logger
+
+	// The prefix for your graphite metric. Defaults to empty string.
+	Prefix string
+}
+
+// Bridge pushes metrics to the configured graphite server.
+type Bridge struct {
+	url      string
+	interval time.Duration
+	prefix   string
+
+	g      prometheus.Gatherer
+	logger log.Logger
+}
+
+// NewBridge returns a pointer to a new Bridge struct.
+func NewBridge(c *Config) (*Bridge, error) {
+	b := &Bridge{}
+
+	if c.URL == "" {
+		return nil, errors.New("graphite bridge: no url given")
```

Comment:
While I can imagine it's useful in certain situations, we haven't put origin hints into the error string anywhere else. Suggestion:
"missing URL"